### PR TITLE
fix(datepicker): restore focus to trigger element

### PR DESCRIPTION
**Pull request by crisbeto**
* The datepicker now restores focus to whatever element was focused before it was open.
* Adds a test for the functionality that closes the datepicker when pressing escape.

### CODE REVIEW

**Comment by mmalerba** [comment]
I think for datepicker we don't want to close on tab (see various accessible datepicker implementations: http://www.webaxe.org/accessible-date-pickers/)

**Comment by crisbeto** [comment]
In that case we'd need to trap focus inside the popup, otherwise it goes out of the page.

**Comment by mmalerba** [comment]
I believe it does: https://github.com/angular/material2/blob/master/src/lib/datepicker/datepicker-content.html#L1

The behavior might be a little weird though, because the dialog (which we use for the touch UI) does its own focus trapping. What I'd like to eventually do is create a `MdPopup`, similar to `MdDialog`. Things like select and auto-complete could take advantage of this as well. One of the APIs on `MdPopup` would allow you to set it to floating mode where its acts more like a dialog (not attached to the input) that way datepicker could just use the popup instead of the overlay/dialog hybrid approach

**Comment by crisbeto** [comment]
It doesn't look like the focus trapping works when it's not in touch mode. Should I remove the tabbing behavior from this PR and try address the focus trapping separately?

**Comment by mmalerba** [comment]
yeah sgtm

File: src/lib/datepicker/datepicker.ts [hunk 1]
```diff
@@ -30,7 +32,7 @@ import {Subscription} from 'rxjs/Subscription';
 import {MdDialogConfig} from '../dialog/dialog-config';
 import {DateAdapter} from '../core/datetime/index';
 import {createMissingDateImplError} from './datepicker-errors';
-import {ESCAPE} from '../core/keyboard/keycodes';
+import {ESCAPE, TAB} from '../core/keyboard/keycodes';
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker.ts]: TAB no longer needed

File: src/lib/datepicker/datepicker.ts [hunk 2]
```diff
@@ -226,6 +229,11 @@ export class MdDatepicker<D> implements OnDestroy {
     if (this._calendarPortal && this._calendarPortal.isAttached) {
       this._calendarPortal.detach();
     }
+    if (this._focusedElementBeforeOpen && 'focus' in this._focusedElementBeforeOpen) {
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker.ts]: why is the `'focus' in` part needed?

Comment (crisbeto) [inline src/lib/datepicker/datepicker.ts]: It's there for IE's sake. In some cases it assigns the `activeElement` to something that doesn't have a `focus` method. It's what was throwing that 4mb error in the CI.

Comment (mmalerba) [inline src/lib/datepicker/datepicker.ts]: interesting, good to know

**Comment by PhilipSultanescu** [comment]
Also got this focus problem, pretty important for us. Can someone please rebase and merge it?

**Comment by PhilipSultanescu** [comment]
Works very good, thank you for merging this.
